File: src/Search/BinarySearch/BinarySearch.ts
/**
 * Returns index of first occurrence of a given value in an array utilizing binary search algorithm.
 *
 * @param {number[]} array - Sorted array.
 * @param {number} searchedValue - Value whose position is being searched.
 * @returns {?number} - Index of the first occurrence of `searchedValue`.
 */
export function binarySearch(array: number[], searchedValue: number) {
  let lower = 0;
  let upper = array.length - 1;

  while (lower <= upper) {
    const splitPosition = Math.floor((lower + upper) / 2);
    if (array[splitPosition] < searchedValue) {
      lower = splitPosition + 1;
    } else if (array[splitPosition] > searchedValue) {
      upper = splitPosition - 1;
    } else {
      return splitPosition;
    }
  }
  return null;
}

File: src/Search/BinarySearch/binary_search.py
def binary_search_iterative(array, searched_value):
    lower = 0
    upper = len(array) - 1

    while lower <= upper:
        middle = (lower + upper) // 2
        if array[middle] < searched_value:
            lower = middle + 1
        elif array[middle] > searched_value:
            upper = middle - 1
        else:
            return middle

    return None


def binary_search_recursive(array, searched_value, lower=None, upper=None):
    if lower is None:
        lower = 0
    if upper is None:
        upper = len(array)

    middle = (lower + upper) // 2

    if lower == upper:
        return None

    if array[middle] == searched_value:
        return middle

    if array[middle] > searched_value:
        return binary_search_recursive(array, searched_value, middle + 1, upper)

    return binary_search_recursive(array, searched_value, lower, middle - 1)
